### PR TITLE
Upstream changes

### PR DESCRIPTION
**Pull request by barker29**
This PR proposes contributing some changes that were used to support another project into the main github smoothG repo.

At the moment several tests are failing, which probably needs to be resolved before this can be merged.

### CODE REVIEW

**Comment by chakshinglee** [comment]
The bug is most likely due to the changes I made in `SparseIdentity` and `SparseDiag`. Specifically, there is some issue when the size is 0. I have made a commit to take care of this special case, please check if it works.

There are other tests that are failing. I think this is because of the way we define `unit_weight` is different. The current definition should be more appropriate, so we should update the test results accordingly.

File: examples/pde.hpp [hunk 1]
```diff
@@ -861,12 +981,15 @@ SPE10Problem::SPE10Problem(const char* permFile, int nDimensions, int spe10_scal
     }
 
     InitGraph();
-    ComputeGraphWeight(unit_weight);
+    // TODO FIXME: unit_weight is the wrong name for what I am doing here
```
Comment (chakshinglee) [inline examples/pde.hpp]: If we want to return edge weights corresponding to the case of constant 1 permeability, then I think the updated approach in this PR is doing the right thing. Note that in this case the edge weight is not 1, but weighted by geometrical quantities like face area, cell center distance, etc.

In the original examples (`sampler`, `ml-sampler`, `ml-sampler4`), we used a pure graph with edge weights being 1 (geometrical quantities are not considered). So we need to update the results if we want to honor the finite volume discretization.

Comment (barker29) [reply examples/pde.hpp]: Based on these comments I removed the option, and rebaselined the `sampler` tests. Now all the tests pass. If you're happy with this and think the rebaselining was legitimate, then I think we're all set.

**Review by chakshinglee — APPROVED** [review]
Your new changes look good to me, and all the test pass now.

**Pull request merged**